Add snapcraft

popey opened the pull request:
Hello!

This pull request adds the capability to build a snap of Akira. I realise you're super busy with your crowdfunding campaign, and packaging isn't your top priority, so no worries if you don't have time to review and/or merge it right now! :D 

I have tested this on Ubuntu 18.04 and KDE Neon. It builds using `snapcraft` and `multipass`. Here's how I build it.

```
snap install snapcraft --classic
snap install multipass --beta --classic
git clone https://github.com/popey/Akira
git checkout "add-snapcraft"
cd Akira
snapcraft
```

This should build a snap around 6.6MB in size. It leverages the `gnome-3-28-1804` content snap to save some space, and reduce duplication. 

Any questions, happy to answer them here. Good luck with the campaign!

Alecaddd commented:
Hi @popey 
Thank you so much for this absolutely wonderful PR.
I'm gonna review it later this evening!

I think I will create a dedicated branch for `snap-packaging` and merge it there since we were using Travis to test it for Huston CI and the elementary OS AppCenter build.

I will probably leave the `master` branch "build-agnostic" and have dedicated branches to build the various packages.

I also create the Akira project on Snapcraft and I will be sure to properly link it.
Thanks again, I'm really excited about this!

bilelmoussaoui commented:
@popey Hey, thanks for your interest on Akira and your PR. I'm not very familiar with Snap, so I can't really review this. 
Can we get this under `build-aux/snap` directory with the build command on a README file there? 
Also, can we ping from time to time to get either updates or whatever needed with snap packaging? 
Thanks!

albfan commented:
@popey I reworked your snap on https://github.com/akiraux/Akira/tree/wip/popey/snap but I get stuck on 

https://travis-ci.org/akiraux/Akira/jobs/502033237#L218

For ubuntu xenial I can only find https://packages.ubuntu.com/xenial/python3-distutils-extra, so not sure how to fix that

If you want you can connect your fork with travis and run only setup for snap. In my branch using a build matrix is fixed

nisrulz commented:
I found out @popey wrote about his process to create the snap. Linking it here, since I think it might be helpful: https://snapcraft.io/blog/zero-to-snap-rev-up-your-packaging

popey commented:
Hello! me again. I've updated the PR to reflect many changes including updates to Akira itself.. I'm now using the snap GNOME extension which makes the snap smaller & faster to start. I've updated the travis config to use lxd - you can't use --destructive mode to build on xenial for a bionic target. LXD will spin up a bionic container in which to build.

bilelmoussaoui commented:
Wasn't there a github action to build a snap? We have moved away from travis to github actions

popey commented:
Ok, I just updated the travis yaml while I was here because it had no hope of building the snap successfully. If someone from the project team registers an 'akira' account on the Snap Store, and visits the store page, they can just add this repo, and the store will start building it on each commit to master. (if this PR lands).

Alecaddd commented:
@popey this is so beautiful I'm almost crying.
I'm setting up everything on the snap store and later today I'll review and merge this PR.
Thank you so much!!!

Alecaddd commented:
@popey sorry to bother you, but the build on Snapcraft is failing.
Here's the last part of the log

```
Configuring config.vala using configuration
src/meson.build:11: WARNING: The variable(s) 'RELEASE_NAME', 'VERSION_INFO' in the input file 'src/config.vala.in' are not present in the given configuration data.
Configuring com.github.akiraux.akira.desktop.in using configuration
WARNING: Gettext not found, all translation targets will be ignored.

data/meson.build:29:0: ERROR: Tried to assign an invalid value to variable.

A full log can be found at /build/akira/parts/akira/build/snapbuild/meson-logs/meson-log.txt
meson --prefix=/usr snapbuild
Failed to run 'meson --prefix=/usr snapbuild' for 'akira': Exited with code 1.
Verify that the part is using the correct parameters and try again.
sed: can't read /build/akira/parts/akira/install/usr/share/applications/com.github.akiraux.akira.desktop: No such file or directory
Build failed
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 266, in run
    self.build()
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 255, in build
    env=env)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 102, in run_build_command
    return self.backend.run(args, env=full_env, **kwargs)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/lxd.py", line 536, in run
    subprocess.check_call(cmd, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['lxc', 'exec', 'lp-bionic-amd64', '--env', 'LANG=C.UTF-8', '--env', 'SHELL=/bin/sh', '--env', 'SNAPCRAFT_BUILD_INFO=1', '--env', 'SNAPCRAFT_IMAGE_INFO={"build-request-id": "lp-55463614", "build-request-timestamp": "2020-03-31T22:50:28Z", "build_url": "https://launchpad.net/~build.snapcraft.io/+snap/d7c8bbf796bc79cad8f2a29c180458f9/+build/888738"}', '--env', 'SNAPCRAFT_BUILD_ENVIRONMENT=host', '--env', 'http_proxy=http://10.10.10.1:8222/', '--env', 'https_proxy=http://10.10.10.1:8222/', '--env', 'GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy', '--', '/bin/sh', '-c', 'cd /build/akira && linux64 snapcraft']' returned non-zero exit status 2
Revoking proxy token...
RUN: /usr/share/launchpad-buildd/bin/in-target scan-for-processes --backend=lxd --series=bionic --arch=amd64 SNAPBUILD-888738
Scanning for processes to kill in build SNAPBUILD-888738
```

It seems that the `config.vala` doesn't get properly generated.

giacomoalbe commented:
As of the recently updated README, you need the gettext dependency,
otherwise meson will complain in that completely unrelated way.

I already tried building Akira on Ubuntu and the problem was solved once I
installed that dependency.

Il giorno mer 1 apr 2020 alle 07:36 Alessandro <notifications@github.com>
ha scritto:

> @popey <https://github.com/popey> sorry to bother you, but the build on
> Snapcraft is failing.
> Here's the last part of the log
>
> Configuring config.vala using configuration
> src/meson.build:11: WARNING: The variable(s) 'RELEASE_NAME', 'VERSION_INFO' in the input file 'src/config.vala.in' are not present in the given configuration data.
> Configuring com.github.akiraux.akira.desktop.in using configuration
> WARNING: Gettext not found, all translation targets will be ignored.
>
> data/meson.build:29:0: ERROR: Tried to assign an invalid value to variable.
>
> A full log can be found at /build/akira/parts/akira/build/snapbuild/meson-logs/meson-log.txt
> meson --prefix=/usr snapbuild
> Failed to run 'meson --prefix=/usr snapbuild' for 'akira': Exited with code 1.
> Verify that the part is using the correct parameters and try again.
> sed: can't read /build/akira/parts/akira/install/usr/share/applications/com.github.akiraux.akira.desktop: No such file or directory
> Build failed
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 266, in run
>     self.build()
>   File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 255, in build
>     env=env)
>   File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 102, in run_build_command
>     return self.backend.run(args, env=full_env, **kwargs)
>   File "/usr/lib/python2.7/dist-packages/lpbuildd/target/lxd.py", line 536, in run
>     subprocess.check_call(cmd, **kwargs)
>   File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
>     raise CalledProcessError(retcode, cmd)
> CalledProcessError: Command '['lxc', 'exec', 'lp-bionic-amd64', '--env', 'LANG=C.UTF-8', '--env', 'SHELL=/bin/sh', '--env', 'SNAPCRAFT_BUILD_INFO=1', '--env', 'SNAPCRAFT_IMAGE_INFO={"build-request-id": "lp-55463614", "build-request-timestamp": "2020-03-31T22:50:28Z", "build_url": "https://launchpad.net/~build.snapcraft.io/+snap/d7c8bbf796bc79cad8f2a29c180458f9/+build/888738"}', '--env', 'SNAPCRAFT_BUILD_ENVIRONMENT=host', '--env', 'http_proxy=http://10.10.10.1:8222/', '--env', 'https_proxy=http://10.10.10.1:8222/', '--env', 'GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy', '--', '/bin/sh', '-c', 'cd /build/akira && linux64 snapcraft']' returned non-zero exit status 2
> Revoking proxy token...
> RUN: /usr/share/launchpad-buildd/bin/in-target scan-for-processes --backend=lxd --series=bionic --arch=amd64 SNAPBUILD-888738
> Scanning for processes to kill in build SNAPBUILD-888738
>
> It seems that the config.vala doesn't get properly generated.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/akiraux/Akira/pull/70#issuecomment-607043622>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYFJPZC57M6KPMLW33RSFDRKLHHLANCNFSM4GSFOHUA>
> .
>

popey commented:
#315 now makes it build okay. It's currently in review in the store as the dbus interface used needs security review. I have requested this on your behalf on the forum thread below.

https://forum.snapcraft.io/t/akira-dbus-auto-connect-request/16346

popey commented:
Ok, dbus name allowed, and it looks like the application is now available in the edge channel. https://snapcraft.io/akira
Adventurous users can install with `snap install akira --edge`.
Feel free to ping me with any further snap-related issues! :D

abienz commented:
Nice it installed very swiftly, I installed it on Arch, I did have a problem with fonts loading unfortunately, see attached.

![AkiraSnap-20200401](https://user-images.githubusercontent.com/888222/78166178-4906c500-7444-11ea-842a-6657582abb66.png)

popey commented:
@abienz Sorry it didn't work for you. There's a long standing fontconfig issue which plagues all containerised apps. There's some issue with ABI incompatibility. Can you please try:

```
sudo rm /var/cache/fontconfig/*
rm ~/.cache/fontconfig/*
```

Then re-launch.

(by the way, is there a slack / irc / mattermost / discord / aol chat thing you guys use I can hang out in?)

Alecaddd commented:
@popey for sure, I sent you an invite to our dev Slack channel, we'd be honoured to have you.
I also created an issue for Sequeler, which I tried in the past to create a snap for, but failed: https://github.com/Alecaddd/sequeler/issues/305

abienz commented:
Thanks clearing the font cache worked a treat.

Alecaddd commented:
@abienz if I'm not wrong, snaps automatically update in the background, so every time we merge something to master, it should spit out a new release.
Bare in mind that the app is still in alpha, and we won't officially publish on the stable channel until all the most basic features are in.
Apologies in advance for any breakage :sweat_smile:

abienz commented:
@Alecaddd Yeah no worries I'm a Patreon and follow the chat here and in the Discord, I know the score.

If anything I'm interested in helping to find bugs for you.
